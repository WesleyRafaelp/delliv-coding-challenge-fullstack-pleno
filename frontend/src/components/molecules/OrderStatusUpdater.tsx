import React, { useEffect, useState } from 'react';
import Button from '../atoms/Button';
import { useDispatch, useSelector } from 'react-redux';
import { getOrders }  from '@/store/reducers/orders';

interface OrderStatusUpdaterProps {
  currentStatus: string;
  onUpdateStatus: (newStatus: string) => void;
  availableStatusOptions: string[];
}

const OrderStatusUpdater: React.FC<OrderStatusUpdaterProps> = ({ currentStatus, onUpdateStatus, availableStatusOptions }) => {
  const [newStatus, setNewStatus] = useState(currentStatus);
  const [isEditing, setIsEditing] = useState(false);
  
  const handleUpdateStatus = () => {
    onUpdateStatus(newStatus);
    setIsEditing(false);
  };

  return (
    <div>
      {isEditing ? (
        <div className="flex items-center">
          <select
            className="border border-gray-300 rounded p-2 mr-2 focus:outline-none focus:ring focus:border-blue-300"
            value={newStatus}
            onChange={(e) => setNewStatus(e.target.value)}
          >
            {availableStatusOptions.map((status) => (
              <option key={status} value={status}>
                {status}
              </option>
            ))}
          </select>
          <Button type='button' onClick={handleUpdateStatus}>Atualizar Status</Button>
        </div>
      ) : (
        <Button type='button' onClick={() => setIsEditing(true)}>Editar Status</Button>
      )}
    </div>
  );
};

export default OrderStatusUpdater;
