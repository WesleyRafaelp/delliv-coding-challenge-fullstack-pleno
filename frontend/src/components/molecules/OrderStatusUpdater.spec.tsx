import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import OrderStatusUpdater from './OrderStatusUpdater';

describe('OrderStatusUpdater Component', () => {
  it('renders OrderStatusUpdater component', () => {
    const onUpdateStatusMock = jest.fn();
    const availableStatusOptions = ['Pending', 'Shipped', 'Delivered'];
    render(
      <OrderStatusUpdater
        currentStatus="Pending"
        onUpdateStatus={onUpdateStatusMock}
        availableStatusOptions={availableStatusOptions}
      />
    );
    
    expect(screen.getByText(/editar status/i)).toBeInTheDocument();
  });

  it('handles status update when "Atualizar Status" button is clicked', () => {
    const onUpdateStatusMock = jest.fn();
    const availableStatusOptions = ['Pending', 'Shipped', 'Delivered'];
    render(
      <OrderStatusUpdater
        currentStatus="Pending"
        onUpdateStatus={onUpdateStatusMock}
        availableStatusOptions={availableStatusOptions}
      />
    );

    fireEvent.click(screen.getByText(/editar status/i));
    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'Shipped' } });
    fireEvent.click(screen.getByText(/atualizar status/i));

    expect(onUpdateStatusMock).toHaveBeenCalledWith('Shipped');
  });

  it('matches snapshot', () => {
    const onUpdateStatusMock = jest.fn();
    const availableStatusOptions = ['Pending', 'Shipped', 'Delivered'];
    const { asFragment } = render(
      <OrderStatusUpdater
        currentStatus="Pending"
        onUpdateStatus={onUpdateStatusMock}
        availableStatusOptions={availableStatusOptions}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
