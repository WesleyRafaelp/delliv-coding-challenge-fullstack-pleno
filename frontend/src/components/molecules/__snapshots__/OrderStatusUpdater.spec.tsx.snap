// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderStatusUpdater Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline-blue active:bg-blue-800"
      type="button"
    >
      Editar Status
    </button>
  </div>
</DocumentFragment>
`;
